---
  html_document:
    toc: true
    highlight: zenburn
    theme: united
---

```{r setup, echo=FALSE}
knitr::opts_chunk$set(tidy=TRUE, highlight=TRUE, dev="png",
               cache=FALSE, highlight=TRUE, autodep=TRUE, warning=FALSE, error=FALSE,
               message=FALSE, prompt=TRUE, comment='', fig.cap='')
```

                                        # Overview
This project is Manuel Leyva in the Geha lab. Questions about this analysis
can be raised by outside investigators by [opening an issue](https://github.com/hbc/geha-il4ko-rnaseq/issues) or by people involved in the project by [posting on Basecamp](https://basecamp.com/1762765/projects/11524447).


```{r qc-setup}
library(ggplot2)
library(reshape)
library(gplots)
library(edgeR)
library(CHBUtils)
library(pheatmap)
library(DESeq2)
library(tximport)
library(logging)
basicConfig()
project_summary = "../project-summary.csv"
counts_file = "../combined.counts"
tx2genes_file = "../tx2gene.csv"
cbPalette <- c("#999999", "#E69F00", "#56B4E9", "#009E73", "#F0E442",
"#0072B2", "#D55E00", "#CC79A7")
summarydata = data.frame(read.table(project_summary, header=TRUE, sep=","), row.names="Name", check.rows=FALSE)
summarydata$Name = rownames(summarydata)
summarydata = summarydata[order(summarydata$Name),]
if (file.exists(tx2genes_file)) {
  loginfo("Using gene counts calculated from the Sailfish transcript counts.")
  sf_files = file.path("..", "..", rownames(summarydata), "sailfish",
                      rownames(summarydata), "quant.sf")
  names(sf_files) = rownames(summarydata)
  tx2gene = read.table(tx2genes_file, sep=",", row.names=NULL, header=FALSE)
  txi.salmon = tximport(sf_files, type="salmon", tx2gene=tx2gene,
                        reader=readr::read_tsv, countsFromAbundance="scaledTPM")
  counts = round(data.frame(txi.salmon$counts, check.names=FALSE))
} else {
  loginfo("Using gene counts calculated from featureCounts.")
  counts = read.table(counts_file, header=TRUE, row.names="id", check.names=FALSE)
}
counts = counts[, order(colnames(counts)), drop=FALSE]
colnames(counts) = gsub(".counts", "", colnames(counts))

# this is a list of all non user-supplied metadata columns that could appear
known_columns = c("Name", "X.GC", "Exonic.Rate", "Sequences.flagged.as.poor.quality",
    "rRNA_rate", "Fragment.Length.Mean", "Intronic.Rate", "Intergenic.Rate",
    "Mapping.Rate", "Quality.format", "Duplication.Rate.of.Mapped", "Mapped",
    "rRNA", "Sequence.length", "Transcripts.Detected", "Mean.Per.Base.Cov.",
    "Genes.Detected", "Unique.Starts.Per.Read", "unique_starts_per_read",
    "complexity", "X5.3.bias", "Duplicates.pct", "Duplicates", "Mapped.reads",
    "Median.insert.size", "Mapped.reads.pct", "Total.reads",
    "avg_coverage_per_region", "Mapped.Reads")
metadata = summarydata[, !colnames(summarydata) %in% known_columns, drop=FALSE]
```

```{r heatmap-function}
get_heatmap_fn = function(summarydata) {
    # return the pheatmap function with or without metadata
    if(ncol(metadata) == 0) {
       return(pheatmap)
    }
    else {
    # rownames(metadata) = summarydata$Name
    heatmap_fn = function(data, ...) {
        pheatmap(data, annotation=metadata, ...)
    }
    return(heatmap_fn)
}}
heatmap_fn = get_heatmap_fn(summarydata)
```

# Quality control metrics

```{r qualimap-run}
qualimap_run = "Mapped" %in% colnames(summarydata)
```

The number of mapped reads looks fine, if a little on the high side. 100 M reads
or so per sample.
## Mapped reads
```{r mapped-plot, eval=qualimap_run}
ggplot(summarydata, aes(x=Name, y=Mapped)) +
    theme_bw(base_size=10) +
    theme(panel.grid.major = element_line(size = .5, color = "grey"),
          axis.text.x = element_text(angle=90)) +
    geom_bar(stat="identity") +
    ylab("mapped reads") + xlab("")
```

```{r mapped-plot-noqualimap, eval=!qualimap_run}
ggplot(summarydata, aes(x=Name, y=Mapped.reads)) +
    theme_bw(base_size=10) +
    theme(panel.grid.major = element_line(size = .5, color = "grey"),
          axis.text.x = element_text(angle=90)) +
    geom_bar(stat="identity") +
    ylab("mapped reads") + xlab("")
```

## Genomic mapping rate
The rate of mapping looks fine as well.
```{r mapping-rate-plot, eval=qualimap_run}
ggplot(summarydata, aes(x=Name, y=Mapping.Rate)) +
    geom_bar(stat="identity") +
    ylab("mapping rate") + xlab("") +
    theme_bw(base_size=10) +
    theme(panel.grid.major = element_line(size = .5, color = "grey"),
          axis.text.x = element_text(angle=90))
```

```{r mapping-rate-plot-noqualimap, eval=!qualimap_run}
ggplot(summarydata, aes(x=Name, y=Mapped.reads.pct)) +
    geom_bar(stat="identity") +
    ylab("mapping rate") + xlab("") +
    theme_bw(base_size=10) +
    theme(panel.grid.major = element_line(size = .5, color = "grey"),
          axis.text.x = element_text(angle=90))
```

## Number of genes detected
The number of genes detected looks fine.
```{r genes-detected-plot}
dd = data.frame(Name=colnames(counts), Genes.Detected = colSums(counts > 0))
ggplot(dd, aes(x=Name, y=Genes.Detected)) +
    geom_bar(stat="identity") +
    theme_bw(base_size=10) +
    theme(panel.grid.major = element_line(size = .5, color = "grey"),
          axis.text.x = element_text(angle=90)) +
    ylab("genes detected") + xlab("")
```

## Exonic mapping rate
The exonic mapping rate looks okay, most of the reads map to exons.
```{r exonic-mapping-plot, eval="Exonic.Rate" %in% colnames(summarydata)}
ggplot(summarydata, aes(x=Name, y=Exonic.Rate)) +
    geom_bar(stat="identity") +
    theme_bw(base_size=10) +
    theme(panel.grid.major = element_line(size = .5, color = "grey"),
          axis.text.x = element_text(angle=90)) +
    ylab("exonic mapping rate") + xlab("")
```

## rRNA mapping rate
The rRNA rate looks fine. It looks like there is more rRNA contamination in
the batch 3 samples, but overall its small in all of the samples. We'd be
worried if it was 20% instead of 2%.

```{r check-rRNA-eval}
eval_rRNA = "rRNA_rate" %in% colnames(summarydata) & !sum(is.na(summarydata$rRNA_rate)) == nrow(summarydata)
```
```{r rRNA-rate-plot, eval=eval_rRNA}
ggplot(summarydata, aes(x=Name, y=rRNA_rate)) +
    geom_bar(stat="identity") +
    theme_bw(base_size=10) +
    theme(panel.grid.major = element_line(size = .5, color = "grey"),
          axis.text.x = element_text(angle=90)) +
    ylab("rRNA rate") + xlab("")
```

## Boxplot of log10 counts per gene
These distributions look okay, you can see a batch effect in batch 3 here.
```{r boxplot-raw}
melted = melt(counts)
colnames(melted) = c("sample", "count")
melted$sample = factor(melted$sample)
melted = melted[order(melted$sample),]
melted$count = log(melted$count)
ggplot(melted, aes(x=sample, y=count)) + geom_boxplot() +
    theme_bw(base_size=10) +
    theme(panel.grid.major = element_line(size = .5, color = "grey"),
          axis.text.x = element_text(angle=90)) + xlab("")
```

## Boxplot of log10 TMM-normalized counts per gene
Trimmed mean of M-values (TMM) normalization is described
[here](http://genomebiology.com/2010/11/3/R25)

Robinson, M. D., & Oshlack, A. (2010). A scaling normalization method for differential expression analysis of RNA-seq data. Genome Biology, 11(3). doi:10.1186/gb-2010-11-3-r25

The batch effect isn't entirely removed due to the normalization, there is a larger
spread in the batch 3 samples.

```{r boxplot-normalized}
y = DGEList(counts=counts)
y = calcNormFactors(y)
normalized_counts = cpm(y, normalized.lib.sizes=TRUE)
melted = melt(normalized_counts)
colnames(melted) = c("gene", "sample", "count")
melted$sample = factor(melted$sample)
melted = melted[order(melted$sample),]
melted$count = log(melted$count)
ggplot(melted, aes(x=sample, y=count)) + geom_boxplot() +
    theme_bw(base_size=10) +
    theme(panel.grid.major = element_line(size = .5, color = "grey"),
          axis.text.x = element_text(angle=90)) + xlab("")
```

## Density of log10 TMM-normalized counts
You can see the batch effect here, on the x-axis are the log counts for
each sample.
```{r density-normalized}
ggplot(melted, aes(x=count, group=sample, color=sample)) +
    geom_density() +
    theme_bw(base_size=10) +
    theme(panel.grid.major = element_line(size = .5, color = "grey"),
          axis.text.x = element_text(angle=90)) + xlab("")
```

## Correlation (Pearson) heatmap of TMM-normalized counts
Here we try clustering the samples using a couple different metrics. In
each attempt, the samples cluster together more by batch, indicating a large
batch effect.
```{r pearson-heatmap-normalized, eval=ncol(counts) > 1}
heatmap_fn(cor(normalized_counts, method="pearson"))
```

## Correlation (Spearman) heatmap of TMM-normalized counts
```{r spearman-heatmap-normalized, eval=ncol(counts) > 1}
heatmap_fn(cor(normalized_counts, method="spearman"))
```

## PCA plot
We can also see the batch effect if we plot the first and second principal
components against each other.
```{r pca, eval=ncol(counts) > 1}
dds = DESeqDataSetFromMatrix(countData=counts, colData=summarydata, design=~Name)
vst = varianceStabilizingTransformation(dds)
plotPCA(vst, intgroup=c("Name"))
```

## summary of QC
The samples look fine, except for the large batch effect. We can see looking
at the PCA plot that 78% of the variation between the samples can be
explained by PC1, which separates on batch. Only 13% of the variation can
be explained by PC2, which seems to separate the cells based on genotype.

So one question is, could there possibly have been a sample swap between
any of the samples? Here were the mappings from the FASTQ files to the
samples that we used:

```{r mapping-files, echo=FALSE}
DT::datatable(read.table("../../geha-rnaseq.csv", sep=",", header=TRUE))
```

Moving forward, we'll assume there was no sample swap and see what the
differences are. We'll do differential expression and look at the differences
between the IL4 knockouts and the wild types by comparing the IL4 KO to its
matched batch mate.

```{r deseq2-tximport-helper-functions}
# snagged from development version of DESeq
DESeqDataSetFromTximport <- function(txi, colData, design, ...)
{
  counts <- round(txi$counts)
  mode(counts) <- "integer"
  dds <- DESeqDataSetFromMatrix(countData=counts, colData=colData, design=design, ...)
  stopifnot(txi$countsFromAbundance %in% c("no","scaledTPM","lengthScaledTPM"))
  if (txi$countsFromAbundance %in% c("scaledTPM","lengthScaledTPM")) {
    message("using length scaled TPM counts from tximport")
  } else {
    message("using counts and average transcript lengths from tximport")
    lengths <- txi$length
    dimnames(lengths) <- dimnames(dds)
    assays(dds)[["avgTxLength"]] <- lengths
  }
  return(dds)
}

subset_tximport = function(txi, rows, columns) {
    txi$counts = txi$counts[rows, columns]
    txi$abundance = txi$abundance[rows, columns]
    txi$length = txi$length[rows, columns]
    return(txi)
}
```

```{r de-setup}
library(DEGreport)
library(vsn)
design = ~batch+genotype
condition = "genotype"
```

# Differential expression

```{r deseq2-expression-analysis, results='asis'}
counts <- counts[rowSums(counts>0)>1,]
if(exists("txi.salmon")) {
   loginfo("Using Sailfish gene counts for the DESeq2 model.")
   txi.salmon = subset_tximport(txi.salmon, rownames(counts), colnames(counts))
   dds = DESeqDataSetFromTximport(txi.salmon, colData=summarydata, design=design)
} else {
  loginfo("Using counts from featureCounts for the DESeq2 model.")
  dds = DESeqDataSetFromMatrix(countData=counts,
      colData=summarydata, design = design  )
}
geoMeans = apply(counts, 1, function(row) if (all(row == 0)) 0 else
                 exp(mean(log(row[row != 0]))))
dds = estimateSizeFactors(dds, geoMeans=geoMeans)
dds = DESeq(dds)
```

## Effect of variance stabilization

```{r deseq-diagnostics, results='asis'}
par(mfrow=c(1,3))
notAllZero <- (rowSums(counts(dds))>0)
rld <- rlog(dds)
vsd <- varianceStabilizingTransformation(dds)
rlogMat <- assay(rld)
vstMat <- assay(vsd)

meanSdPlot(log2(counts(dds,normalized=TRUE)[notAllZero,] + 1))
meanSdPlot(assay(rld[notAllZero,]))
meanSdPlot(assay(vsd[notAllZero,]))
```

## Dispersion estimates

```{r dispersion-estimate}
plotDispEsts(dds)
```

```{r deseq2-handler}
handle_deseq2 = function(dds, summarydata, column) {
  all_combs = combn(levels(summarydata[,column]), 2, simplify=FALSE)
  all_results = list()
  contrast_strings = list()
  for(comb in all_combs) {
    contrast_string = paste(comb, collapse=" vs ")
    contrast = c(column, comb)
    res = results(dds, contrast=contrast)
    res = res[order(res$padj),]
    all_results = c(all_results, res)
    contrast_strings = c(contrast_strings, contrast_string)
  }
  names(all_results) = contrast_strings
  return(all_results)
}
```

## MA-plots

```{r DESeq-output, results='asis'}
all_results = handle_deseq2(dds, summarydata, condition)
len = length(all_results)
nr = ceiling( len / 3 )
nc = ceiling( len / nr )
par(mfrow=c(nr,nc))
for(i in seq(length(all_results))) {
  plotMA(all_results[[i]])
  title(paste("MA plot for contrast", names(all_results)[i]))
}
```

## Volcano-plots

```{r DESeq-volcano}
for(i in seq(length(all_results))) {
  stats = as.data.frame(all_results[[i]][,c(2,6)])
  p = volcano_density_plot(stats, title=names(all_results)[i], lfc.cutoff=1.5)
  print(p)
}
```

## DEGreport

```{r get-groups}
get_groups <- function(d, comp, condition)
{
  g <- unlist(strsplit(comp," "))
  g1 <- d$Name[d[, (names(d)==condition)]==g[1]]
  g2 <- d$Name[d[, (names(d)==condition)]==g[3]]
  list(g1,g2)
}
```

### Pvalues-vs-Mean

Here we plot some information about how the p-values are correlated with the
mean or the standard deviation.

```{r DEGreport-M}
plots = list()
scale_factor = round(1/nr * 14)
for(i in seq(length(all_results))) {
  plots[[i]] = degMean(all_results[[i]]$pvalue, rlogMat) +
  theme_bw(base_size = scale_factor) +
  ggtitle(paste0("Pvalues-vs-Mean for ", names(all_results)[i]))
}
do.call(grid.arrange,plots)
```

### Pvalues-vs-Variation

```{r DEGreport-V}
plots = list()
for(i in seq(length(all_results))) {
  plots[[i]] = degVar(all_results[[i]]$pvalue, rlogMat) +
  theme_bw(base_size = scale_factor) +
  ggtitle(paste0("Pvalues-vs-Variation for ", names(all_results)[i]))
}
do.call(grid.arrange,plots)
```

### Mean-vs-Variation
```{r DEGreport-MV}
plots = list()
for(i in seq(length(all_results))) {
  g <- get_groups(summarydata, names(all_results)[i], condition)
  if(length(g[[1]]) < 2 | length(g[[2]]) < 2) {
     next
   }
  plots[[i]] = degMV(g[[1]], g[[2]], all_results[[i]]$pvalue, counts(dds,normalized=TRUE)) +
  theme_bw(base_size = scale_factor) +
  ggtitle(paste0("Mean-vs-Variation for ", names(all_results)[i]))
}
if(length(plots) > 0) {
    do.call(grid.arrange,plots)
}
```

## Differentially expressed genes

```{r DESeq-tables, results='asis'}
for(i in seq(length(all_results))) {
  cat(paste("Lowest adjusted p-value hits for", names(all_results)[i]))
  out_df = as.data.frame(all_results[[i]])
  write.table(out_df, file=paste(names(all_results)[i], ".tsv", sep=""),
                         quote=FALSE, sep="\t", row.names=TRUE, col.names=TRUE)
  cat("\n")
}
```

```{r lookup-biomart-stuff}
library(biomaRt)
mouse = useMart(biomart = "ENSEMBL_MART_ENSEMBL",
                dataset="mmusculus_gene_ensembl",
                host = "jul2015.archive.ensembl.org")
conversions = getBM(attributes=c("ensembl_gene_id", "mgi_symbol", "gene_biotype"),
                    mart=mouse)
rrna_biotypes = c("rRNA", "Mt_rRNA", "misc_RNA", "snRNA", "snoRNA",
                    "tRNA", "Mt_tRNA")
rrna_genes = unique(subset(conversions, gene_biotype %in% rrna_biotypes)$ensembl_gene_id)
```

```{r annotate-with-symbol}
library(dplyr)
res = data.frame(all_results[[1]])
res$id = rownames(res)
res = res %>% left_join(conversions, by=c("id"="ensembl_gene_id"))
```

Here we show all genes with padj < 0.05, and output a file with all of the
results that can be loaded into Excel and filtered however you like.

`baseMean` is the mean count value across all samples, `log2FoldChange` is
the log 2 fold change between the WT and IL4KO samples
(positive is higher in the IL4KO), `lfcSE` is the log 2 fold change standard
error, `stat` is the value of the test statistic, `pvalue` is the pvalue and
`padj` is the BH-adjusted pvalue. If you want to filter for significant hits,
use `padj` since that is multiple-hypothesis corrected.

```{r display-and-write}
res = res[, c("id", colnames(res)[colnames(res) != "id"])]
sig_genes = subset(res, padj < 0.05)
DT::datatable(sig_genes)
write.table(res, file=paste(names(all_results)[1], ".tsv", sep=""),
                        quote=FALSE, sep="\t", row.names=FALSE, col.names=TRUE)
```

It looks like a lot of the top hits are immunoglobulin genes, and they appear
down, which I think should be an expected result. So it looks like maybe
there wasn't as sample swap after all. There are `r nrow(subset(res, padj < 0.05))`
genes called differentially expressed using an adjusted p-value cutoff of 0.05.
Here is the breakdown of the significant hits by biotype:

```{r by-biotype-table}
biotype_table = as.data.frame(table(subset(res, padj < 0.05)$gene_biotype))
colnames(biotype_table) = c("biotype", "significant genes")
DT::datatable(biotype_table)
```

While there are a bunch of miRNA in the differential gene list, we can't really
call miRNA differences from this data, since we
didn't do a small RNA experiment. What we are likely picking up is unspliced pre-miRNA. Picking up more pre-miRNA might indicate there is more of the mature miRNA being expressed. I talked to Lorena Pantano, our miRNA expert in the core and she said there isn't
a clear correlation between pre-miRNA expression and mature miRNA expression. If
you wanted to do anything with the miRNA results,
you'd have to follow up with qPCR or a small RNA sequencing experiment or something to look at the mature miRNA if you wanted to make that into a publishable finding.

## GO/KEGG analysis
We need two pieces of information to do the GO and KEGG analysis. The first is
the background set of genes that are expressed in this sample. We'll call
that background set any gene with a baseMean expression of at least 10. The
second is the list of genes we are calling differentially expressed and
also have an expression of at least 10. Then we test to see if there are pathways
overrepresented in the differentially expressed list compared to the
background list.

```{r write-expressed-gene-identifiers}
expressed = subset(res, baseMean > 10)
write.table(expressed$id, file="expressed.txt", col.names=FALSE,
            row.names=FALSE, quote=FALSE)
```

```{r write-de-identifiers}
de = subset(res, baseMean > 10 & padj < 0.05)
write.table(de$id, file="de-nonswap.txt", col.names=FALSE,
            row.names=FALSE, quote=FALSE)
```

Now we load those into [WebGestalt](http://bioinfo.vanderbilt.edu/webgestalt/)
to look at GO and KEGG enrichment. We looked at only GO ontology terms and
pathways that had at least 5 terms, and called a hit significant if it had a BH
adjusted p-value < 0.05.

[GO ontology](images/GO-noswap.gif)
[KEGG analysis](web/KEGG-noswap.html)

## sample swap
Manuel and I talked over the phone, and we think there might have been a
sample swap after all. Il13ra1 should be down in the Il4KO cells because Il13ra1 is
controlled by Il4. That doesn't seem to be the case though:

```{r Il13ra1-counts, results='asis'}
knitr::kable(counts["ENSMUSG00000017057",])
```

It looks from that that batch 1 is the IL4KO and batch 3 is the WT. This
would also make the PCA plot make a lot more sense.

He is going to check with the core and his notes to see if they can figure out
if a swap happened. He is also going to do some qPCR and look at some other
markers to see if they can identify other genes that might be markers of a
swap. Here is a table of the mapping of the FASTQ files to the
sample metadata that we used for reference.

```{r mapping-files-redone, echo=FALSE}
knitr::kable(read.table("../../geha-rnaseq.csv", sep=",", header=TRUE))
```

In the meantime, we'll look at the effect of batch, which we think is might
actually be the genotype. `batch1` should be IL4ko and `batch3` is the WT.

```{r doctor-sampletable}
swapped_results = handle_deseq2(dds, summarydata, "batch")
swapped_res = data.frame(swapped_results[[1]])
swapped_res$id = rownames(swapped_res)
swapped_res = swapped_res %>% left_join(conversions, by=c("id"="ensembl_gene_id"))
write.table(swapped_res, file=paste("swapped-", names(all_results)[1], ".tsv", sep=""),
                        quote=FALSE, sep="\t", row.names=FALSE, col.names=TRUE)
```

```{r DESeq-volcano-swapped}
for(i in seq(length(swapped_results))) {
  stats = as.data.frame(swapped_results[[i]][,c(2,6)])
  p = volcano_density_plot(stats, title=names(swapped_results)[i], lfc.cutoff=1.5)
  print(p)
}
```

There are `r nrow(subset(swapped_res, padj < 0.05))`  genes which are different
between `batch1` and `batch3`, so a huge difference.

```{r write-swapped-de-identifiers}
swapped_de = subset(swapped_res, baseMean > 10 & padj < 0.05)
write.table(swapped_de$id, file="de-swap.txt", col.names=FALSE,
            row.names=FALSE, quote=FALSE)
```

```{r write-counts}
out_count_table = counts
out_count_table$id = rownames(out_count_table)
out_cols = colnames(out_count_table)
out_cols = c("id", out_cols[out_cols != "id"])
out_count_table = out_count_table[, out_cols]
write.table(out_count_table, row.names=FALSE, col.names=TRUE, quote=FALSE, sep="\t",
            file="counts.txt")
```

Here are the GO and KEGG results for that:

[GO ontology](images/GO-swap.gif)
[KEGG analysis](web/KEGG-swap.html)
